### PR TITLE
Add subset sampling support for distributed settings.

### PR DESCRIPTION
**Pull request by Miladrzh**
This is useful for training on a subset of data or splitting validation
sets. The subset sampling only exists for non-distributed settings in
torch.utils.data.sampler.SubsetRandomSampler before this code.

### Description

I have added subset sampling support to the 'torch.utils.data.distributed.DistributedSampler' class. Before this code, subset sampling was only possible in non-distributed settings. 

This is useful for training on a subset of data or splitting validation sets. Since the DataLoader class only accept one sampler, this code is necessary for splitting data in distributed settings (among GPUs and using a subset of data rather than the whole data). 

### Issue

I did not submit any issues. But I needed this code for splitting my training data into the validation set. It was easy in non-distributed settings with 'torch.utils.data.sampler.SubsetRandomSampler'. But the only option for distributed settings was 'torch.utils.data.distributed.DistributedSampler', which is not supporting subsets. 

### Testing

The code was inherited from a tested class that existed before. The rest of the code is tested with training a model on cifar100. I also tested different corner cases to make sure that the code always has expected behavior and in case of bad usage, it has asserts and provides explanatory exceptions. 

cc @SsnL @VitalyFedyunin @ejguan @NivekT

### CODE REVIEW

**Comment by Miladrzh** [comment]
@rohan-varma Hi, Any updates, sir?

**Comment by Miladrzh** [comment]
@rohan-varma 
Any updates, sir? Is there any feedback for me to at least learn something ;)?

**Comment by kit1980** [comment]
@rohan-varma Please take a look if this is something we want to do.

**Review by rohan-varma — COMMENTED** [review]
Hi @Miladrzh, thanks for your contribution! This sounds reasonable overall, but would need some unittests to get landed. 

You can check example distributed sampler tests: https://github.com/pytorch/pytorch/blob/master/test/test_dataloader.py and contributing guide: https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md to get started with writing unittests.

**Comment by Miladrzh** [comment]
@rohan-varma 
Thanks. I have added all similar tests to the `distributed sampler` class.
I also add an additional test to check if the indices generated are within the subset that was specified in the beginning.

**Comment by kit1980** [comment]
@Miladrzh can you fix the lint issues?
